src/fam_struct_wrapper.rs

```diff
@@ -0,0 +1,590 @@
+            if len > T::max_len() {
+                return Err(Error::SizeLimitExceeded);
+            }
+            // Reserve additional capacity.
+            self.reserve(additional_elements as usize);
+        }
+
+        // Perform the actual len update.
+        self.len = len;
+        self.as_mut_fam_struct().set_len(len);
+        let required_mem_allocator_len = FamStructWrapper::<T>::mem_allocator_len(len);
+        // This is safe since enough capacity has been reserved
+        unsafe {
+            self.mem_allocator.set_len(required_mem_allocator_len);
+        }
+
+        // If the len needs to be decreased:
+        if additional_elements < 0 {
+            // Deallocate unnecessary memory
+            self.mem_allocator.shrink_to_fit();
+            // Update capacity
+            self.capacity = FamStructWrapper::<T>::fam_len(self.mem_allocator.capacity());
+        }
+
+        Ok(())
+    }
+
+    /// Propagate the [`FamStruct`](trait.FamStruct.html) state
+    /// to the [`FamStructWrapper`](struct.FamStructWrapper.html).
+    /// This is needed after `self.as_ptr()` is supplied as param to an FFI call
+    /// which may change it's state.
+    ///
+    pub fn sync(&mut self) -> Result<(), Error> {
+        let new_len = self.as_fam_struct_ref().len();
+        self.set_len(new_len)
+    }
+
+    /// Appends an element to the end of the collection and updates `len`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The element that will be appended to the end of the collection.
+    ///
+    /// # Error: When len is already equal to max possible len it returns Error::SizeLimitExceeded
+    ///
+    pub fn push(&mut self, entry: T::Entry) -> Result<(), Error> {
+        let new_len = self.len + 1;
+        self.set_len(new_len)?;
+        self.as_mut_slice()[new_len - 1] = entry;
+
+        Ok(())
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// # Arguments
+    ///
+    /// * `f` - The function used to evaluate whether an entry will be kept or not.
+    ///         When `f` returns `true` the entry is kept.
+    ///
+    pub fn retain<P>(&mut self, mut f: P)
+    where
+        P: FnMut(&T::Entry) -> bool,
+    {
+        let mut num_kept_entries = 0;
+        {
+            let entries = self.as_mut_slice();
+            for entry_idx in 0..entries.len() {
+                let keep = f(&entries[entry_idx]);
+                if keep {
+                    entries[num_kept_entries] = entries[entry_idx];
+                    num_kept_entries += 1;
+                }
+            }
+        }
+
+        // This is safe since this method is not increasing the len
+        self.set_len(num_kept_entries).expect("invalid length");
+    }
+}
+
+impl<T: Default + FamStruct> PartialEq for FamStructWrapper<T> {
+    fn eq(&self, other: &FamStructWrapper<T>) -> bool {
+        self.len == other.len && self.as_slice() == other.as_slice()
```

Comment:
Sorry I know I asked this regarding the old PR as well, but I don't remember the conclusion :( Is it a valid occurence to have `self.as_slice() == other.as_slice()`, but `self.len != other.len`?